### PR TITLE
Stabilise exhaustive patterns feature

### LINKED ISSUE #104034: `#[feature(exhaustive_patterns)]` thinks irrefutable pattern is refutable when enum contains struct with private fields.
The following code compiles on nightly ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=55ac489a6a9fcd1189f52d4d2c254e94)):

```rust
#![feature(exhaustive_patterns, never_type)]

enum Either<A, B> {
    A(A),
    B(Wrapper<B>),
}

struct Wrapper<T>(T);

fn foo() -> Either<(), !> {
    Either::A(())
}

fn main() {
    let Either::A(()) = foo();
}
```

But if the wrapper type is moved into a module to make the field private, then it does not: ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=81d51fa51530ad29918e0d6e078c3070)):

```rust
#![feature(exhaustive_patterns, never_type)]

mod inner {
    pub struct Wrapper<T>(T);
}

enum Either<A, B> {
    A(A),
    B(inner::Wrapper<B>),
}

fn foo() -> Either<(), !> {
    Either::A(())
}

fn main() {
    let Either::A(()) = foo();
}
```

```
error[E0005]: refutable pattern in local binding: `Either::B(_)` not covered
  --> src/main.rs:17:9
   |
17 |     let Either::A(()) = foo();
   |         ^^^^^^^^^^^^^ pattern `Either::B(_)` not covered
   |
   = note: `let` bindings require an "irrefutable pattern", like a `struct` or an `enum` with only one variant
   = note: for more information, visit https://doc.rust-lang.org/book/ch18-02-refutability.html
note: `Either<(), !>` defined here
  --> src/main.rs:9:5
   |
7  | enum Either<A, B> {
   |      ------
8  |     A(A),
9  |     B(inner::Wrapper<B>),
   |     ^ not covered
   = note: the matched value is of type `Either<(), !>`
help: you might want to use `if let` to ignore the variant that isn't matched
   |
17 |     if let Either::A(()) = foo() { todo!() }
   |     ++                           ~~~~~~~~~~~
```

Making the field `pub` makes it compile again.

Is this expected? I suppose I could understand if it is; the private fields could change to make `Wrapper<!>` actually be constructable and therefore make the pattern refutable. But if that's the case, then I think the compiler should point this out, like `note: the pattern is currently irrefutable, but the type contains private fields which may change in the future to make the pattern refutable`.

If indeed this is expected, is there any way for me as a library author to somehow convince the compiler that I will never change the fields to make the type constructable? I want users of my library to be able to elide match arms when I give them an `Either<A, !>`.

@rustbot label +T-compiler +F-never_type +D-confusing +requires-nightly +S-bug-has-mcve

### PR DESCRIPTION
**Pull request by cjgillot**
# Stabilisation report

This feature allows to omit visibly uninhabited patterns from a match.

```rust
fn main() {
  let x: Result<u8, !> = Ok(5);
  match x {
    Ok(y) => y,
    // no need to add a `Err` pattern, as `!` is uninhabited.
  }
}
```

This omission is only possible when the types are *visibly* uninhabited at the use point, ie. only if the uninhabitedness comes from a field that is accessible. As a drive-by this PR adds a note explaining this design choice.
Fixes https://github.com/rust-lang/rust/issues/104034

The relevant tests are in `tests/ui/pattern`.

I am not aware of any unresolved questions.

### CODE REVIEW

**Comment by fmease** [comment]
> I am not aware of any unresolved questions.

Not sure if this still relevant but as far as I am aware the concern raised in http://smallcultfollowing.com/babysteps/blog/2018/08/13/never-patterns-exhaustive-matching-and-uninhabited-types-oh-my/ hasn't been addressed yet.

**Comment by oli-obk** [comment]
I would like to evaluate how this feature interacts with constants used in patterns. I don't expect any issues, but I want to make sure we're not adding more edge cases to the const pattern logic

**Comment by cjgillot** [comment]
> > I am not aware of any unresolved questions.
> 
> Not sure if it still relevant but as far as I am aware the concern raised in http://smallcultfollowing.com/babysteps/blog/2018/08/13/never-patterns-exhaustive-matching-and-uninhabited-types-oh-my/ hasn't been addressed.

The rules around UB have matured since that blog post has been written, so I'm not sure where the concern lies any more. @nikomatsakis does this concern still stand?

> I would like to evaluate how this feature interacts with constants used in patterns. I don't expect any issues, but I want to make sure we're not adding more edge cases to the const pattern logic

The logic here is purely type-based, so there should not be any interaction. What do you have in mind?

**Comment by oli-obk** [comment]
> The logic here is purely type-based, so there should not be any interaction. What do you have in mind?

Ah I guess that makes sense. The pattern gets rejected as a required pattern just on its own. I guess the only thing i'll check is what happens to an `Err(const {panic!()})` pattern. It should still get evaluated and fail compilation.

**Comment by varkor** [comment]
Could you leave a comment on the [tracking issue](https://github.com/rust-lang/rust/issues/51085) responding to the blocking issues mentioned in https://github.com/rust-lang/rust/issues/51085#issuecomment-743807861?

cc @Nadrieril

**Comment by nikomatsakis** [comment]
@cjgillot 

> The rules around UB have matured since that blog post has been written

As far as I know, the questions raised in that post are still relevant and have not been resolved. @rust-lang/opsem is probably better positioned to judge whether this stabilization makes those questions any easier or harder. I can try to bring that stuff back into cache if they don't weigh in. =)

**Comment by RalfJung** [comment]
Right, specifically the interesting questions involve pointer indirections: given `x: &!`, does this PR let one write `match x {}`? Operationally speaking, where exactly does the UB come from if that piece of code is ever executed?

Similar with `x: *const !` and `match *x {}`: conceptually this match merely creates a place without ever looking at it (it is the match arms that might do the "looking", but there are no arms here), so it is far from clear where UB would come from here. That's why the blog post proposes `match *x { ! }`, where the `!` is the match arm that makes the "looking" part explicit and hence explains where the UB comes from.

**Comment by cjgillot** [comment]
It references are not considered uninhabited, so the 'x: &!' case is not allowed. 

As for the "what does the looking", MIR building introduces a fake read before the unreachable terminator for precisely this reason. So the '!' arm is the whitespace between the last explicit arm and the closing brace. 

https://rust-lang.zulipchat.com/#narrow/stream/136281-t-opsem/topic/Soundness.20of.20.60exhaustive_patterns.60 might be an easier place to discuss this.

**Comment by RalfJung** [comment]
> It references are not considered uninhabited, so the 'x: &!' case is not allowed.

Oh, okay. That should be documented clearly and covered by tests to ensure this will not be changed with some drive-by PR in the future.

If inhabitedness never recurses below pointer indirections, so at least on first thought I think this does not have the issues that were discussed in the blog post.

> As for the "what does the looking", MIR building introduces a fake read before the unreachable terminator for precisely this reason. So the '!' arm is the whitespace between the last explicit arm and the closing brace.

When and where exactly such a fake read is introduced is not entirely clear though. E.g. arguably `match *x { _ => {} }` should *not* have a 'fake read' and should only assert that the place itself exists.

**Comment by Nadrieril** [comment]
> references are not considered uninhabited, so the 'x: &!' case is not allowed.

Well, `&!` is technically considered inhabited atm, but pattern-matching usually looks into references. E.g. as a user I expect that matching on `x: &Result<T, !>` and `x.as_ref(): Result<&T, &!>` should have the same number of arms.

> arguably `match *x { _ => {} }` should _not_ have a 'fake read' and should only assert that the place itself exists.

With `exhaustive_patterns`, the branch in `match *x { _ => {} }` will be warned to be unreachable. If both `match *x {}` and `match *x { _ => {} }` make sense but do something different, that's unfortunate.

> conceptually this match merely creates a place without ever looking at it

How do we make sense of "creating a place for a value of type `!`? Layout-wise there can be no such thing right?

**Comment by cjgillot** [comment]
For `&!`, I'm not sure I understand why this is tricky. If we were to declare `&!` as uninhabited, then matching on a `&!` would omit the arm because the `&!` is uninhabited, without trying to dereference it. The uninhabitedness would come from the impossibility to create a `&!`, wouldn't it? But that's more of a theoretical point, and out of the scope of this PR.

For now, this test exists: https://github.com/rust-lang/rust/blob/master/tests/ui/pattern/usefulness/always-inhabited-union-ref.rs

This does not apply to `*const !` which is inhabited by dangling pointers.

> When and where exactly such a fake read is introduced is not entirely clear though. E.g. arguably `match *x { _ => {} }` should _not_ have a 'fake read' and should only assert that the place itself exists.

We currently emit a fakeread of the scrutinee in all cases. https://github.com/rust-lang/rust/pull/103208 would restrict the fakeread to the branch with an `unreachable` terminator only.

**Comment by Nadrieril** [comment]
> That is definitely the case: The former generates an `Unreachable` terminator, the latter is a NOP.

Then we need to fix the exhaustiveness checker to accept both somehow. Presumably `match *x { _ => {} }` should be marked as unreachable unless around unsafe code, right?

> Why not? It's a regular ZST, layout-wise. In fact eventually we probably want `addr_of!(*ptr)` on `ptr: *const !` to be allowed, so constructing such a place can then happen in well-defined programs.

That's unsettling. I would expect that for consistency we need `!` to behave pretty differently than a ZST. Intuitively the only possible `*const !` should be the null pointer or something. I imagine generic code might force us to treat it like a ZST anyway? Are we sure that's consistent?

**Comment by asquared31415** [comment]
All `*const !` must be valid to *have*, since we say that you can make whatever pointers you want safely, it's only operations on them that might be unsafe.  `500 as *const !` is safe code, so we can't say that only the null pointer is valid for `*const !`.

File: src/tools/rust-analyzer/crates/hir-ty/src/diagnostics/match_check/deconstruct_pat.rs [hunk 1]
```diff
@@ -41,6 +41,9 @@
 //! or-patterns; instead we just try the alternatives one-by-one. For details on splitting
 //! wildcards, see [`SplitWildcard`]; for integer ranges, see [`SplitIntRange`].
 
+// This uses uninhabited types to encode statically unused variants.
+#![allow(unreachable_patterns)]
```
Comment (matklad) [inline src/tools/rust-analyzer/crates/hir-ty/src/diagnostics/match_check/deconstruct_pat.rs:45]: Drive by comment, but the fact that changing rustc requires atomically changing `rust-analyzer` doesn't feel right to me. Ideally, this change is done by ra devs in ra repo when ra upgrades to the new rust version. 

Perhaps rust-lang/rust build system should build rust-analyzer with `--cap-lints allow`? From rustc point of view, rust-analyzer pretty much is a dependency which might not have upgraded to the latest compiler yet.

**Comment by Nadrieril** [comment]
`let _ = *x` is UB if the pointer is unaligned, right? I think it would be most consistent if the alignment of `!` is infinity, i.e. the only aligned `*const !` is the null pointer. This is compatible with `addr_of` not being UB and `500 as *const !` being safe.

This means the arm in `match *x { _ => {} }` is unreachable. By the normal validity rules for `&T` (ref must be non-null and aligned), this also implies that there is no valid `&!`. All of this makes my life as the main author of the exhaustiveness algorithm much easier :D

**Comment by asquared31415** [comment]
I think that any proposal that changes the semantics of `&!` is a non-starter for stability reasons.  The `!` type is accessible on stable (just by a different name), it can be observed that `!: Sized`, and we guarantee that `&T` is laid out the same as `usize` for all `T: Sized`.  To change this for `&!` specifically would require removing the `Sized` trait (not something that can be done) and *also* adding a whole new system to the compiler to support uninhabited references, and some way to make it ineligible to be used as the `T` in a `fn uwu<T: ?Sized>()` because it also would behave differently to other fat references.

This can affect *safe code* as well, it's not just unsafe code and weird UB.

**Comment by Nadrieril** [comment]
This doesn't require any changes around `Sized` or the byte representation of references. I'm merely reusing the existing requirements on values of type `&T` (namely that they are non-null and aligned to the alignement of `T`) and declaring that uninhabited types have an alignment of 0. It's similar to how converting an unaligned pointer to a reference is UB even if we never dereference it.

I might be missing something but I think this doesn't affect safe code. There never was and never will be a safe way to create a `&!` reference. Even unsafe code is ok because `&T` must point to initialized memory, and there's no way to have initialized memory to an uninhabited type. The one thing that changes is the behavior of `*const T`. I fully expect unsafe code will have done unholy things with that. Also pretty sure lots of code assumes that alignments are non-zero so that would break a lot of stuff.

Wait if what I just said is true then it is already the case that `&!` has no valid values since there cannot be initialized memory for `!`. Unless we treat it weirdly right now?

**Comment by RalfJung** [comment]
> `let _ = *x` is UB if the pointer is unaligned, right?

As I said above, `addr_of!(*x)` ever being UB at all is considered surprising by many. This includes alignment requirements. This means the `*x` place expression needs to evaluate successfully even for unaligned pointers. So no, I think `let _ = *x;` is *not* UB if the pointer is unaligned. Or at least I don't want to luck us into a corner where it should be UB, this is something @rust-lang/opsem still has to decide on.

**Comment by Nadrieril** [comment]
> This means the *x place expression needs to evaluate successfully even for unaligned pointers

ohh I see I didn't realize that could change

> From the perspective of exhaustiveness checking, what is the issue with match *x { _ => {} } not being UB?

I changed my mind, it's fine, see below

> we shouldn't preempt that discussion

ok, so let's go for the conservative take. I think that means `&!` is inhabited,  and `match *x { _ => {} }` is reachable. That feels forward-compatible?

Here's my take on the situation: before we can think of stabilizing this feature, we (I) need to clarify the interaction between validity and pattern-matching. The feature is currently implemented wrong: among other things it doesn't know that `match *x { _ => {} }` is reachable for `x: *const !`. More generally I haven't handled this carefully even for non-empty types. So step 1 is I need to chat with opsem and fix that.

Then remain open two questions of ergonomics that feel important.

One problem is that `match *x {}` now asserts validity of the value. @nikomatsakis argued that was maybe too implicit and we want to do something clever about it. Afaik we haven't made progress on that since the never patterns blog post. Doing nothing is forwards-compatible with adding warnings later, but is a potential UB footgun.

Other problem is that, from what I currently understand, both
```rust
match x {
	Ok(_) => {}
}
```
and
```rust
match x {
	Ok(_) => {}
	Err(_) => {}
}
```
make sense. The difference is that the second one doesn't assert the validity of the contained `!`. This is definitely surprising to users, and only makes sense around unsafe code. Something should be done, maybe change the behavior around unsafe code? Someone tell me what to do.

In conclusion, I think the feature is not ready for stabilization

**Comment by nikomatsakis** [comment]
I don't want to let perfect by enemy of good here. Am I correct that It is backwards compatible to stabilize something that requires `_` and then later make it unnecessary -- we already have this for empty enum types, after all. I mean backwards compatible here the narrow sense of: all code that used to compile still compiles without changes in the semantics (modulo deny lints), though it may get new lint warnings about dead code. In the worst case we could leverage an edition.

In other words, I favor stabilization if it is simply preserving the "status quo" where `&U` (where `U` is an uninhabited type) is considered inhabited. I also favor eventually finding some way to make `let x: &U; match x { }` compile, I just don't know how to go about it yet.

**Comment by JakobDegen** [comment]
Agreed that I don't see any reason that that isn't backwards compatible

**Comment by RalfJung** [comment]
Seeing the open question around `match *x {}`, maybe we should for now stabilize uninhabitedness-aware `match` in a way that
- `&T` is always considered inhabited
- we don't apply this when the matched-on place involves a `Deref` projection

If there is no `Deref` projection then there must already have been UB earlier in the code so what we do around `match` matter much less. In contrast, with `match *x {}` the UB is at the match.

**Comment by CAD97** [comment]
I concur that stabilization currently seems like it needs to be limited to by-move matches which unambiguously only use the by-move binding mode on an expiring temporary, where match arms of `_ => unreachable!()` and `_x => unreachable!()` have the same behavior.

Alternatively, perhaps stabilize only the `let Ok(x) = x;` form? Since in that form it's not possible to have an `_` arm... except that's exactly what `let .. else` is, so perhaps it's even more egregious there? Since explaining a "desugar" as `else { nonbehavior!() }` is actively incorrect.

<details><summary>As a side note,</summary>

There's no intrinsic rule that `let _ = $place;` and `match $place { _ => () };` *must* have identical semantics. It's probably a poor idea, but we could keep the `let` as a no-op and make the `match` assert byte validity of the scrutinee if we really wanted to. As a prior illustration of divergent behavior, consider that `if` guards are permitted with `match` but not `let`, and that said `if` guards might require different retagging of the scrutinee than otherwise to prohibit the utilization of shared mutability on the matched pattern. (It's complicated.)

Generally speaking, people tend to find the behavior of `_` patterns surprising, since most scrutinees tend to be by-reference or temporaries, and partial-by-move is relatively uncommon, especially now with default binding modes meaning it's more likely to match `&(Copy, Move)` as `&(ref Copy, ref _)` than `&(move Copy, move _)`. It wouldn't be *completely* unthinkable to change a `_` pattern to capture the scrutinee in the default binding mode and immediately drop it, introducing a different way to get an "ignore binding mode" (perhaps a bare `ref`).

I've probably written code which isn't *wrong* but is unintentionally suboptimal, where I `match` some variable binding and by-move it in some arms, but `_` it in a fallback arm, where the intent is to consume the value and drop it / free its resources, but in the `_` case it's actually not been moved from and will only drop at the end of scope. It's essentially the exact same surprise/footgun as the temporary lifetime of locks, just at a bigger scale. [[constructed example]](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9e0f94386e7a45d541f89d8ea6945774)

To elide uninhabited patterns, the compiler is effectively inserting `Pattern(!) => nonbehavior!()` arms... but the issue is that this is inserting new (primitive pattern matching) AM ops which can cause fresh UB before the nonbehavior, but which aren't present in the source and which wouldn't be present with a simpler `_ => nonbehavior!()` arm. This nearly brings me back to suggesting that nonraw retagging should assert byte validity of the reborrowed memory, such that `&!` becomes UB to acquire, even though it's still technically inhabited, but basically any mention of the reference as a reference results in reborrowing it.

</details>

File: tests/ui/pattern/usefulness/empty-match.exhaustive_patterns.stderr
Comment (jhpratt) [inline tests/ui/pattern/usefulness/empty-match.exhaustive_patterns.stderr:1]: This file should be deleted, as `empty-match.rs` has the revisions removed, correct?

**Review by jhpratt — COMMENTED** [review]
Other than this one question, the diff LGTM.

**Comment by JakobDegen** [comment]
In addition to the example I just shared in #103208 , @RalfJung's claim that

> If there is no Deref projection then there must already have been UB earlier

also seems wrong in at least one other case:

```rust
union U {
    a: (),
    b: !,
}

let x: U = U { a: () };

match x.b {
    _ if {
        println!("Executed match!");
        loop {}
    } => (),
}
```

This also [compiles](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8b9806f22cddbb331de258d8be9c76dc) today and is presumably not UB.

I don't understand this PR well enough yet to know if this is a problem.

**Comment by RalfJung** [comment]
Ah right, I forgot about union fields.

**Comment by cjgillot** [comment]
> In addition to the example I just shared in #103208 , @RalfJung's claim that
> 
> > If there is no Deref projection then there must already have been UB earlier
> 
> also seems wrong in at least one other case:
> 
> ```rust
> union U {
>     a: (),
>     b: !,
> }
> 
> let x: U = U { a: () };
> 
> match x.b {
>     _ if {
>         println!("Executed match!");
>         loop {}
>     } => (),
> }
> ```
> 
> This also [compiles](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8b9806f22cddbb331de258d8be9c76dc) today and is presumably not UB.
> 
> I don't understand this PR well enough yet to know if this is a problem.

@JakobDegen This PR does not change the behaviour of that code, but introduces a spurious "unreachable pattern" warning on the `_`.

**Comment by goffrie** [comment]
> This PR does not change the behaviour of that code, but introduces a spurious "unreachable pattern" warning on the `_`.

It looks like even on stable today, there is already an "unreachable pattern" warning in that case, per the playground link?

**Comment by Nyabinary** [comment]
> > This PR does not change the behaviour of that code, but introduces a spurious "unreachable pattern" warning on the `_`.
> 
> It looks like even on stable today, there is already an "unreachable pattern" warning in that case, per the playground link?

Wait, what do you mean by that?

**Comment by goffrie** [comment]
[This playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8b9806f22cddbb331de258d8be9c76dc) shows a warning even on stable:
```
warning: unreachable pattern
  --> src/main.rs:13:13
   |
13 |             _ if {
   |             ^
   |
   = note: `#[warn(unreachable_patterns)]` on by default
```

**Comment by Nadrieril** [comment]
This is indeed a bug, which is tracked in https://github.com/rust-lang/rust/issues/117119. I am about to submit a PR to fix it

**Comment by lcnr** [comment]
Going through all my github assignments rn. Since this PR has been opened `feature(min_exhaustive_patterns)` has been stabilized in #122792. I would expect that stabilizing the rest of `feature(exhaustive_patterns)` requires future discussion and a more involved stabilization proposal. Please open a new PR for that in this case, closing.
